File: src/main/java/org/cip4/xjdf/json/openapi/LocalElement.java
package org.cip4.xjdf.json.openapi;

import lombok.RequiredArgsConstructor;
import org.cip4.xjdf.json.openapi.model.Model;
import org.cip4.xjdf.json.openapi.model.Schema;
import org.w3c.dom.Node;

@RequiredArgsConstructor
public class LocalElement implements Modelable {

    private final Node node;
    private final Context context;

    public boolean isRequired() {
        return !Integer.valueOf(0).equals(getMinOccurs());
    }

    private Integer getMinOccurs() {
        Node minOccursNode = node.getAttributes().getNamedItem("minOccurs");
        if (minOccursNode == null) {
            return null;
        }
        return Integer.parseInt(minOccursNode.getNodeValue());
    }

    private Integer getMaxOccurs() {
        Node maxOccursNode = node.getAttributes().getNamedItem("maxOccurs");
        if (maxOccursNode == null) {
            return null;
        }
        String value = maxOccursNode.getNodeValue();
        if ("unbounded".equals(value)) {
            return Integer.MAX_VALUE;
        } else {
            return Integer.parseInt(value);
        }
    }

    public String getName() {
        Node nameNode = node.getAttributes().getNamedItem("name");
        if (nameNode != null) {
            return nameNode.getNodeValue();
        }
        Node refNode = node.getAttributes().getNamedItem("ref");
        if (refNode != null) {
            return refNode.getNodeValue();
        }
        throw new IllegalStateException("Neither 'name' nor 'ref' attributes are present.");
    }

    private Schema getType() {
        Node typeNode = node.getAttributes().getNamedItem("type");
        if (typeNode == null) {
            return null;
        }
        return context.getNameTranslator().translate(typeNode.getNodeValue());
    }

    private String getRef() {
        Node refNode = node.getAttributes().getNamedItem("ref");
        return refNode != null ? refNode.getNodeValue() : null;
    }

    @Override
    public Model getModel() {
        Schema schema = deriveType();

        Integer maxOccurs = getMaxOccurs();
        if (maxOccurs != null && maxOccurs > 1) {
            schema = listOf(schema);
            Integer minOccurs = getMinOccurs();
            if (minOccurs != null && minOccurs != 0) {
                schema.minItems(minOccurs);
            }
            if (!maxOccurs.equals(Integer.MAX_VALUE)) {
                schema.maxItems(maxOccurs);
            }
        }

        return new Model(getName(), schema);
    }

    private Schema listOf(Schema schema) {
        return new Schema()
            .type("array")
            .items(schema);
    }

    private Schema deriveType() {
        String ref = getRef();
        if (ref != null) {
            return context.getNameTranslator().translate(ref);
        }
        Schema type = getType();
        return type != null ? type : new Schema();
    }
}

File: src/main/kotlin/org/cip4/xjdf/openapi/LocalElement.kt
/*
 * The CIP4 Software License, Version 1.0
 *
 * Copyright (c) 2001-2021 The International Cooperation for the Integration of
 * Processes in Prepress, Press and Postpress (CIP4). All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without modification,
 * are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation and/or
 * other materials provided with the distribution.
 *
 * 3. The end-user documentation included with the redistribution, if any, must
 * include the following acknowledgment: &quot;This product includes software developed
 * by the The International Cooperation for the Integration of Processes in
 * Prepress, Press and Postpress (www.cip4.org)&quot; Alternately, this acknowledgment
 * may appear in the software itself, if and wherever such third-party
 * acknowledgments normally appear.
 *
 * 4. The names &quot;CIP4&quot; and &quot;The International Cooperation for the Integration of
 * Processes in Prepress, Press and Postpress&quot; must not be used to endorse or
 * promote products derived from this software without prior written permission.
 * For written permission, please contact info@cip4.org
 *
 * 5. Products derived from this software may not be called &quot;CIP4&quot;, nor may &quot;CIP4&quot;
 * appear in their name, without prior written permission of the CIP4 organization
 *
 * &quot;CIP4&quot; and &quot;The International Cooperation for the Integration of Processes in
 * Prepress, Press and Postpress&quot; are trademarks of The International Cooperation
 * for the Integration of Processes in Prepress, Press and Postpress (CIP4).  All
 * other names and brands are the property of their respective owners.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESSED OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
 * FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. CIP4 AND ITS CONTRIBUTORS MAKE
 * NO REPRESENTATION THAT THE USE OF THIS SOFTWARE OR DATA WILL NOT INFRINGE ANY
 * PATENTS, COPYRIGHTS, TRADEMARKS, OR OTHER RIGHTS. IN NO EVENT SHALL CIP4 OR ITS
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
 * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
 * OF SUCH DAMAGE.
 * ====================================================================
 *
 * This software consists of voluntary contributions made by many individuals on
 * behalf of the The International Cooperation for the Integration of Processes in
 * Prepress, Press and Postpress and was originally based on software some of which
 * contributed by and copyrighted by Adobe Systems and Man Roland. copyright (c)
 * 2000-2001, Adobe Systems copyright (c) 2001-2001, Man Roland
 *
 * For more information on The International Cooperation for the Integration of
 * Processes in Prepress, Press and Postpress , please see &lt;http://www.cip4.org/&gt;.
 */

package org.cip4.xjdf.openapi

import org.cip4.xjdf.openapi.model.NamedSchema
import org.cip4.xjdf.openapi.model.Schema
import org.w3c.dom.Node

class LocalElement(
    private val node: Node,
    private val context: Context

) {

    val isRequired: Boolean
        get() = minOccurs > 0

    private val minOccurs: Int
        get() = node.attributes.getNamedItem("minOccurs")?.nodeValue
            ?.let { Integer.parseInt(it) }
            ?: 1

    private val maxOccurs: Int
        get() = node.attributes.getNamedItem("maxOccurs")?.nodeValue
            ?.let {
                when (it) {
                    "unbounded" -> null
                    else -> Integer.parseInt(it)
                }
            }
            ?: 1

    internal val name: String
        get() = node.attributes.getNamedItem("name")?.nodeValue
            ?: node.attributes.getNamedItem("ref")?.nodeValue!!

    private val type: Schema?
        get() = node.attributes.getNamedItem("type")?.nodeValue?.let { context.nameTranslator.translate(it) }

    private val ref: Schema?
        get() = node.attributes.getNamedItem("ref")?.nodeValue?.let { Schema.ref(it) }

    fun getModel(): NamedSchema {
        var schema = ref ?: type ?: Schema()
        if (maxOccurs > 1) {
            schema = listOf(schema)
        }
        return NamedSchema(name, schema)
    }

    private fun listOf(schema: Schema): Schema {
        return Schema(
            type = "array",
            items = schema
        )
    }

}